src/main/java/com/oath/halodb/CompactionManager.java

```diff
@@ -306,18 +310,33 @@ private boolean isRecordFresh(IndexFileEntry entry, InMemoryIndexMetaData metaDa
         }
 
         private void rollOverCurrentWriteFile(int recordSize) throws IOException {
-            if (currentWriteFile == null ||  currentWriteFileOffset + recordSize > dbInternal.options.getMaxFileSize()) {
-                if (currentWriteFile != null) {
-                    currentWriteFile.flushToDisk();
-                    currentWriteFile.getIndexFile().flushToDisk();
+            // Make sure there's no unnecessary roll over of the write
+            compactionFileRollOverLock.lock();
+            try {
+                if (currentWriteFile == null || currentWriteFileOffset + recordSize > dbInternal.options
+                    .getMaxFileSize()) {
+                    forceRolloverCurrentWriteFile();
                 }
-                currentWriteFile = dbInternal.createHaloDBFile(HaloDBFile.FileType.COMPACTED_FILE);
-                dbInternal.getDbDirectory().syncMetaData();
-                currentWriteFileOffset = 0;
+            } finally {
+                compactionFileRollOverLock.unlock();
             }
         }
     }
 
+    void forceRolloverCurrentWriteFile() throws IOException {
+        compactionFileRollOverLock.lock();
+        try {
+            if (currentWriteFile != null) {
+                currentWriteFile.flushToDisk();
+                currentWriteFile.getIndexFile().flushToDisk();
+            }
+            currentWriteFile = dbInternal.createHaloDBFile(HaloDBFile.FileType.COMPACTED_FILE);
+            dbInternal.getDbDirectory().syncMetaData();
+            currentWriteFileOffset = 0;
```

Comment:
```currentWriteFileOffset``` is used by compaction thread all the time. When you do the rollover, it will impact ongoing writes. At least this variable need to be ```volatile```.

Considering user not often do snapshot, can we pause compaction job during taking snapshot?